### ISSUE TITLE
Webpack's UglifyJsPlugin causing problems.

### ISSUE THREAD

**Issue by birkir** [mobxjs/mobx-react#127]
I am having problems with the package when compiling for production. I get the following error:
```bash
s.default.createClass is not a function
```
It is originating from https://github.com/mobxjs/mobx-react/blob/master/src/inject.js#L8 where `React` is uglified to `s.default`.

I can resolve it by either:
1. Build the `mobx-react` package without the **UglifyJsPlugin** in `webpack.config.js` and use it directly in my project.
2. Or use the npm version and build my project without the **UglifyJsPlugin** in `webpack.config.js`.

Questions:
1. Is uglifying the build really necessary? 
2. I cannot reproduce the problem with a dead simple project, as something happens when the code size or dependency tree gets to X size.

**Comment by birkir** [comment]
Ok I've been looking better into this issue. To fix the build, I've added the following to my UglifyJs config:

```js
client.plugins.push(new webpack.optimize.UglifyJsPlugin({
  compress: {
    warnings: true,
  },
  mangle: {
    except: ['r'], // I'm guess r is require.
  },
}));
```

But this is only temp solution because the next version that get's published could generate other variable name for `require`.

Looking into other packages (like `mobx` for example), they don't ship their build uglified. Some packages do it as well but strictly name them `package.min.js`. Also you should check out how many issues the uglifyjs package has.

Thank you.

**Comment by birkir** [comment]
I want to stress this issue again.

Check out the components that are part of the `mobx-react` or `mobx-react-devtools` packages:

![image](https://cloud.githubusercontent.com/assets/180773/18845540/93489d36-8411-11e6-8ef8-9a0badef98e1.png)

This is in development mode. The build phase really needs modification asap.

**Comment by mweststrate** [comment]
@birkir can you share your babel setup (I now it has some plugins for alternative handling of default imports etc)? which babel version?

Anybody having the same issue?

**Comment by birkir** [comment]
Sure, I'm working on the ueno. starter kit: https://github.com/ueno-llc/starter-kit

**Comment by mweststrate** [comment]
@birkir especially your first problem is very weird, this setup is here for half a year and used by many people. Even though uglified require calls to React should be resolved by your webpack setup. Your build setup is quite complicated, would you be able to boil it down to a minimal build setup?

I guess your mangle setup should exclude `require`, not `r`? Does the problem still occur without uglify on your site?

@andykog any ideas?

**Comment by andykog** [comment]
Reproduced the issue with @birkir's starter-kit. Will find out why this happens.
Untill that pushed a quick fix: https://github.com/mobxjs/mobx-react/commit/e16633ee2f9a017d6236ee80fc3c62dc129dd7d4
Though, I don't like minifying code by default as well. Strange that uglify.js doesn't complain about being fed with compiled code and so the performance is suffering.

**Comment by birkir** [comment]
Thanks @andykog.

Well, I guess the reason why nobody has encountered this issue within the last half a year is because it happens randomly based on UglifyJs's tree stack and that few people have been using it in production with uglifyjs.

This happens with create-react-app from facebook as well when you add a few components and dependencies.

**Comment by mweststrate** [comment]
@andykog @birkir yeah actually I can't think of a good to minify it as well, it only makes sense if people use it in production from a CDN. Do people still do that nowadays? @mattruby do you use mobx-react from cdn in production or bundle it?

**Comment by mattruby** [comment]
I bundle everything using jawr.  I'll use the cdn in examples but not in production.

**Comment by mweststrate** [comment]
Removed minification from build process and released as `3.5.7`

**Comment by chrisrichard** [comment]
I'm having the same issue when trying to minify a bundle that uses mobx-react. Try to figure out what's going on. In my case, s (which should contain React) is:
```
{ default: { default: 3 } }
```

Maybe an issue with UglifyJS?
